src/ipa_pc/mod.rs

```diff
@@ -176,26 +209,158 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
         }
 
         end_timer!(check_time);
         Some(check_poly)
     }
 
+    /// Perform the succinct check of proof, returning the succinct check polynomial (the xi_s)
+    /// and the GFinal.
+    fn succinct_batch_check_individual_opening_challenges<'a, R: RngCore>(
+        vk: &VerifierKey<G>,
+        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Commitment<G>>>,
+        query_set: &QuerySet<G::ScalarField>,
+        values: &Evaluations<G::ScalarField>,
+        batch_proof: &BatchProof<G>,
+        opening_challenges: &dyn Fn(u64) -> G::ScalarField,
+        _rng: &mut R,
+    ) -> Result<(SuccinctCheckPolynomial<G::ScalarField>, G), Error>
+    {
+        let commitments: Vec<&'a LabeledCommitment<Commitment<G>>> = commitments.into_iter().map(|comm| comm).collect();
+
+        let batch_check_time = start_timer!(|| "Multi poly multi point batch check: succinct part");
+        let evals_time = start_timer!(|| "Compute batched poly value");
+
+        // v_i values
+        let mut v_values = vec![];
+
+        // y_i values
+        let mut y_values = vec![];
+
+        // x_i values
+        let mut points = vec![];
+
+        for (label, (_point_label, point)) in query_set.iter() {
+
+            let y_i = values
+                .get(&(label.clone(), *point))
+                .ok_or(Error::MissingEvaluation {
+                    label: label.to_string(),
+                })?;
+            let v_i = batch_proof.batch_values
+                .get(&label.clone())
+                .ok_or(Error::MissingEvaluation {
+                    label: label.to_string(),
+                })?;
+            v_values.push(*v_i);
+            y_values.push(*y_i);
+            points.push(point);
+        }
+
+        // Commitment of the h(X) polynomial
+        let batch_commitment = batch_proof.batch_commitment.clone();
+
+        // TODO: When we will move to Sponge-based construction, we will absorb the
+        //       batch_commitment only and then squeeze a new challenge
+        // Fresh random challenge x
+        let point = Self::compute_random_oracle_challenge(
+            &to_bytes![
+                batch_commitment,
+                points
+            ]
+                .unwrap(),
+        );
+
+        let mut opening_challenge_counter = 0;
+
+        // lambda
+        let mut cur_challenge = opening_challenges(opening_challenge_counter);
+        opening_challenge_counter += 1;
+
+        let mut computed_batch_v = G::ScalarField::zero();
+
+        for ((&v_i, y_i), x_i) in v_values.iter().zip(y_values).zip(points) {
+
+            computed_batch_v = computed_batch_v + &(cur_challenge * &((v_i - &y_i) / &(point - x_i)));
+
+            cur_challenge = opening_challenges(opening_challenge_counter);
+            opening_challenge_counter += 1;
+        }
+
+        let mut batch_values = vec![];
+        for commitment in commitments.iter() {
```

Comment:
Aren't the values v_i at the fresh challenge point x already collected in v_values?
What is the reason for this extra gathering?

Reply:
These vectors are different. In short v_values contains duplicated values, but batch_values contains only unique values.

In details:

We have vector y_values which contains all polynomials computed in all points. Each value in y_values has a corresponded value in v_values:

For example:

Let assume x is fresh random challenge

Then vectors are computed like this:

y_values[0] = p_1(a), v_values[0] = p_1(x)
y_values[1] = p_1(b), v_values[1] = p_1(x)
y_values[2] = p_2(a), v_values[2] = p_2(x)

as you can see v_values[0] and v_values[1] here has the same values

In batch_values we don't need duplicates, so batch_values for this example will contain only 2 values:

batch_values[0] = p_1(x)
batch_values[1] = p_2(x)